File: Clone App/source/launcher/config_crypto.py
"""
Config Encryption Module - Clone App ULTRA
==========================================
Mã hóa file cấu hình bằng AES-256-CBC.
Key sinh từ HWID của máy → mỗi máy có key riêng.

Author: Clone App Team
Phase: 08 - Security Hardening
"""

import os
import json
import hashlib
import base64
import subprocess
from pathlib import Path
from typing import Optional, Tuple
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend

# Constants
APPDATA_DIR = Path(os.environ.get("APPDATA", "")) / "RunCloneApp"
CONFIG_FILE = APPDATA_DIR / "config.enc"
LEGACY_CONFIG_NAME = "created_users.json"


def run_hidden(cmd: list[str], **kwargs):
    if os.name == "nt" and "creationflags" not in kwargs:
        kwargs["creationflags"] = subprocess.CREATE_NO_WINDOW
    return subprocess.run(cmd, **kwargs)


def get_machine_hwid() -> str:
    """
    Lấy HWID của máy từ nhiều nguồn để tạo unique key.
    Kết hợp: Machine GUID + Processor ID + Volume Serial
    """
    parts = []

    # 1. Machine GUID (từ Registry)
    try:
        result = run_hidden(
            [
                "reg",
                "query",
                "HKLM\\SOFTWARE\\Microsoft\\Cryptography",
                "/v",
                "MachineGuid",
            ],
            capture_output=True,
            text=True,
            timeout=5,
        )
        for line in result.stdout.splitlines():
            if "MachineGuid" in line:
                parts.append(line.split()[-1])
                break
    except Exception:
        pass

    # 2. Computer Name (fallback)
    parts.append(os.environ.get("COMPUTERNAME", "UNKNOWN"))

    # 3. Username (extra entropy)
    parts.append(os.environ.get("USERNAME", "user"))

    combined = "-".join(parts)
    return combined


def derive_key_from_hwid(hwid: str) -> bytes:
    """
    Sinh AES-256 key từ HWID.
    HWID → SHA256 → 32 bytes key
    """
    # Thêm salt cố định để tăng entropy
    salt = b"CloneAppULTRA_SecureConfig_v1"
    data = hwid.encode("utf-8") + salt

    # SHA-256 cho ra 32 bytes = 256 bits (đúng cho AES-256)
    key = hashlib.sha256(data).digest()
    return key


def get_encryption_key() -> bytes:
    """Lấy key mã hóa dựa trên HWID của máy."""
    hwid = get_machine_hwid()
    return derive_key_from_hwid(hwid)


def encrypt_data(data: dict) -> bytes:
    """
    Mã hóa dictionary thành bytes.
    Format: [16 bytes IV] + [Encrypted data]
    """
    key = get_encryption_key()

    # Tạo random IV (Initialization Vector)
    iv = os.urandom(16)

    # Chuyển dict sang JSON string
    plaintext = json.dumps(data, ensure_ascii=False, indent=2).encode("utf-8")

    # Padding to block size (16 bytes for AES)
    padding_length = 16 - (len(plaintext) % 16)
    plaintext += bytes([padding_length]) * padding_length

    # Encrypt với AES-256-CBC
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()

    # Kết hợp IV + ciphertext
    return iv + ciphertext


def decrypt_data(encrypted: bytes) -> Optional[dict]:
    """
    Giải mã bytes thành dictionary.
    """
    if len(encrypted) < 32:  # Minimum: 16 IV + 16 data block
        return None

    key = get_encryption_key()

    # Tách IV và ciphertext
    iv = encrypted[:16]
    ciphertext = encrypted[16:]

    try:
        # Decrypt
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        plaintext = decryptor.update(ciphertext) + decryptor.finalize()

        # Remove PKCS7 padding
        padding_length = plaintext[-1]
        if padding_length > 16:
            return None
        plaintext = plaintext[:-padding_length]

        # Parse JSON
        return json.loads(plaintext.decode("utf-8"))
    except Exception:
        return None


def ensure_config_dir() -> Path:
    """Tạo thư mục config nếu chưa có."""
    APPDATA_DIR.mkdir(parents=True, exist_ok=True)
    return APPDATA_DIR


def find_legacy_config() -> Optional[Path]:
    """
    Tìm file config cũ (plaintext) để migrate.
    Tìm trong thư mục launcher cũ.
    """
    # Thử tìm trong thư mục script đang chạy
    script_dir = Path(__file__).resolve().parent
    legacy_path = script_dir / LEGACY_CONFIG_NAME

    if legacy_path.exists():
        return legacy_path

    return None


def load_config() -> dict:
    """
    Đọc config (tự động migrate từ plaintext nếu cần).

    Priority:
    1. Đọc từ file encrypted mới
    2. Nếu không có → Migrate từ file cũ
    3. Nếu cũng không có → Trả về config rỗng
    """
    ensure_config_dir()

    # 1. Thử đọc file encrypted mới
    if CONFIG_FILE.exists():
        try:
            encrypted = CONFIG_FILE.read_bytes()
            data = decrypt_data(encrypted)
            if data is not None:
                return data
        except Exception:
            pass

    # 2. Thử migrate từ file cũ
    legacy_path = find_legacy_config()
    if legacy_path and legacy_path.exists():
        try:
            data = json.loads(legacy_path.read_text(encoding="utf-8"))

            # Migrate: Save encrypted version
            save_config(data)

            # Xóa file cũ sau khi migrate thành công
            try:
                legacy_path.unlink()
            except Exception:
                pass

            return data
        except Exception:
            pass

    # 3. Trả về config rỗng
    return {"users": [], "apps": []}


def save_config(data: dict) -> bool:
    """
    Lưu config (đã mã hóa).
    """
    try:
        ensure_config_dir()
        encrypted = encrypt_data(data)
        CONFIG_FILE.write_bytes(encrypted)
        return True
    except Exception:
        return False


def get_config_path() -> Path:
    """Trả về đường dẫn file config hiện tại."""
    return CONFIG_FILE


# Export public functions
__all__ = [
    "load_config",
    "save_config",
    "get_config_path",
    "get_machine_hwid",
]

File: Clone App/source/ui/configCrypto.js
/**
 * Config Encryption Module - Clone App ULTRA (Electron)
 * =====================================================
 * Đọc/ghi config đã mã hóa bằng AES-256-CBC.
 * Key sinh từ HWID của máy → mỗi máy có key riêng.
 * 
 * Phase: 08 - Security Hardening
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const os = require('os');

// Constants
const APPDATA_DIR = path.join(process.env.APPDATA || '', 'RunCloneApp');
const CONFIG_FILE = path.join(APPDATA_DIR, 'config.enc');
const LEGACY_CONFIG_NAME = 'created_users.json';
const SALT = Buffer.from('CloneAppULTRA_SecureConfig_v1');

/**
 * Lấy HWID của máy từ nhiều nguồn để tạo unique key.
 * PHẢI KHỚP với Python: get_machine_hwid()
 */
function getMachineHwid() {
    const parts = [];

    // 1. Machine GUID (từ Registry)
    try {
        const result = execSync(
            'reg query "HKLM\\SOFTWARE\\Microsoft\\Cryptography" /v MachineGuid',
            { encoding: 'utf-8', timeout: 5000, windowsHide: true }
        );
        const match = result.match(/MachineGuid\s+REG_SZ\s+(\S+)/i);
        if (match) {
            parts.push(match[1]);
        }
    } catch (e) {
        // Ignore
    }

    // 2. Computer Name (fallback)
    parts.push(process.env.COMPUTERNAME || os.hostname() || 'UNKNOWN');

    // 3. Username (extra entropy)
    parts.push(process.env.USERNAME || os.userInfo().username || 'user');

    return parts.join('-');
}

/**
 * Sinh AES-256 key từ HWID.
 * PHẢI KHỚP với Python: derive_key_from_hwid()
 */
function deriveKeyFromHwid(hwid) {
    const data = Buffer.concat([Buffer.from(hwid, 'utf-8'), SALT]);
    return crypto.createHash('sha256').update(data).digest();
}

/**
 * Lấy key mã hóa dựa trên HWID của máy.
 */
function getEncryptionKey() {
    const hwid = getMachineHwid();
    return deriveKeyFromHwid(hwid);
}

/**
 * Mã hóa dictionary thành bytes.
 * Format: [16 bytes IV] + [Encrypted data with PKCS7 padding]
 */
function encryptData(data) {
    const key = getEncryptionKey();
    const iv = crypto.randomBytes(16);

    // Chuyển object sang JSON
    let plaintext = Buffer.from(JSON.stringify(data, null, 2), 'utf-8');

    // PKCS7 padding
    const paddingLength = 16 - (plaintext.length % 16);
    const padding = Buffer.alloc(paddingLength, paddingLength);
    plaintext = Buffer.concat([plaintext, padding]);

    // Encrypt với AES-256-CBC
    const cipher = crypto.createCipheriv('aes-256-cbc', key, iv);
    cipher.setAutoPadding(false); // Đã tự padding
    const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()]);

    return Buffer.concat([iv, encrypted]);
}

/**
 * Giải mã bytes thành dictionary.
 */
function decryptData(encrypted) {
    if (encrypted.length < 32) {
        return null;
    }

    const key = getEncryptionKey();
    const iv = encrypted.slice(0, 16);
    const ciphertext = encrypted.slice(16);

    try {
        const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
        decipher.setAutoPadding(false);
        let plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);

        // Remove PKCS7 padding
        const paddingLength = plaintext[plaintext.length - 1];
        if (paddingLength > 16 || paddingLength === 0) {
            return null;
        }
        plaintext = plaintext.slice(0, -paddingLength);

        return JSON.parse(plaintext.toString('utf-8'));
    } catch (e) {
        console.error('[ConfigCrypto] Decrypt error:', e.message);
        return null;
    }
}

/**
 * Tạo thư mục config nếu chưa có.
 */
function ensureConfigDir() {
    if (!fs.existsSync(APPDATA_DIR)) {
        fs.mkdirSync(APPDATA_DIR, { recursive: true });
    }
    return APPDATA_DIR;
}

/**
 * Tìm file config cũ (plaintext) để migrate.
 */
function findLegacyConfig(projectRoot) {
    const legacyPath = path.join(projectRoot, 'launcher', LEGACY_CONFIG_NAME);
    if (fs.existsSync(legacyPath)) {
        return legacyPath;
    }
    return null;
}

/**
 * Đọc config (tự động migrate từ plaintext nếu cần).
 */
function loadConfig(projectRoot) {
    ensureConfigDir();

    // 1. Thử đọc file encrypted mới
    if (fs.existsSync(CONFIG_FILE)) {
        try {
            const encrypted = fs.readFileSync(CONFIG_FILE);
            const data = decryptData(encrypted);
            if (data !== null) {
                return data;
            }
        } catch (e) {
            console.error('[ConfigCrypto] Error reading encrypted config:', e.message);
        }
    }

    // 2. Thử migrate từ file cũ
    const legacyPath = findLegacyConfig(projectRoot);
    if (legacyPath) {
        try {
            const rawData = fs.readFileSync(legacyPath, 'utf-8');
            const data = JSON.parse(rawData);

            // Migrate: Save encrypted version
            saveConfig(data);

            // Xóa file cũ sau khi migrate thành công
            try {
                fs.unlinkSync(legacyPath);
                console.log('[ConfigCrypto] Migrated and removed legacy config');
            } catch (e) {
                // Ignore delete error
            }

            return data;
        } catch (e) {
            console.error('[ConfigCrypto] Error migrating legacy config:', e.message);
        }
    }

    // 3. Trả về config rỗng
    return { users: [], apps: [] };
}

/**
 * Lưu config (đã mã hóa).
 */
function saveConfig(data) {
    try {
        ensureConfigDir();
        const encrypted = encryptData(data);
        fs.writeFileSync(CONFIG_FILE, encrypted);
        return true;
    } catch (e) {
        console.error('[ConfigCrypto] Error saving config:', e.message);
        return false;
    }
}

/**
 * Trả về đường dẫn file config hiện tại.
 */
function getConfigPath() {
    return CONFIG_FILE;
}

module.exports = {
    loadConfig,
    saveConfig,
    getConfigPath,
    getMachineHwid,
    APPDATA_DIR,
    CONFIG_FILE
};
